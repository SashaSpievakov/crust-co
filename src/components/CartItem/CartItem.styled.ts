import styled from 'styled-components';

const Article = styled.article`
  display: flex;
  width: 100%;
  position: relative;
  border-top: 1px solid ${({ theme }) => theme.colors.gray};
  padding-top: 30px;
  margin-top: 30px;

  @media screen and (max-width: 850px) {
    flex-direction: column;
    row-gap: 10px;
    align-items: center;
  }
`;

const ImgWrapper = styled.div`
  display: flex;
  align-items: center;
  margin-right: 15px;
  width: 10%;

  img {
    border-radius: 20px;
  }

  @media screen and (max-width: 850px) {
    width: auto;
    margin-right: 0;

    img {
      width: 200px;
      height: 200px;
    }
  }
`;

const InfoWrapper = styled.section`
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 40%;

  h3 {
    font-weight: bold;
    font-size: 22px;
    line-height: 27px;
    letter-spacing: 0.01em;
  }

  p {
    font-size: 18px;
    color: #8d8d8d;
  }

  @media screen and (max-width: 850px) {
    width: auto;
    text-align: center;
  }
`;

const Counter = styled.section`
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 13%;

  span {
    font-size: 20px;
  }

  @media screen and (max-width: 850px) {
    width: auto;
    gap: 11px;
  }
`;

const CountIcon = styled.i`
  stroke-width: 1;
`;

const Price = styled.h3`
  font-size: 20px;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 33%;
`;

const Remove = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-end;
  width: 4%;

  @media screen and (max-width: 850px) {
    position: absolute;
    right: 0;
    top: 20px;
  }
`;

export { Article, Counter, CountIcon, ImgWrapper, InfoWrapper, Price, Remove };
