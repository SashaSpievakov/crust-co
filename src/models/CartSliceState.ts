import { ICartItem } from './ICartItem';

export interface CartSliceState {
  totalPrice: number;
  itemsCount: number;
  items: ICartItem[];
}
