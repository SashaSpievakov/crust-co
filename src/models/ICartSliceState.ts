import { ICartItem } from './ICartItem';

export interface ICartSliceState {
  totalPrice: number;
  itemsCount: number;
  items: ICartItem[];
}
