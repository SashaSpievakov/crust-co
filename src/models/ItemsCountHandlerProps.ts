export interface ItemsCountHandlerProps {
  id: string;
  name: string;
  price: number;
  activeSize: number;
  activeType: number;
  typeNames: string[];
  isFullScreen?: boolean;
}
