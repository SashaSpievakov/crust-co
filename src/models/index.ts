export type { ICartItem } from './ICartItem';
export type { IPizzaItem } from './IPizzaItem';
