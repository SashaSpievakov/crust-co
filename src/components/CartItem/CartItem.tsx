import { memo } from 'react';
import { BsDashLg, BsPlusLg } from 'react-icons/bs';

import { addItem, removeItem, removeItems } from '../../store/slices/cartSlice';
import { ButtonCircle } from '../../styles/Buttons.styled';
import {
  Section,
  ImgWrapper,
  InfoWrapper,
  Counter,
  CountIcon,
  Price,
  Remove,
} from './CartItem.styled';
import { ICartItem } from '../../models/ICartItem';
import { useAppDispatch } from '../../hooks/reduxHooks';

const CartItem = memo(({ id, name, price, size, type, count }: ICartItem) => {
  const dispatch = useAppDispatch();

  const onClickRemove = () => {
    const item: ICartItem = {
      id,
      name,
      price,
      size,
      type,
      count,
    };
    dispatch(removeItem(item));
  };

  const onClickAdd = () => {
    const item: ICartItem = {
      id,
      name,
      price,
      size,
      type,
      count,
    };
    dispatch(addItem(item));
  };

  const onClickDelete = () => {
    const item: ICartItem = {
      id,
      name,
      price,
      size,
      type,
      count,
    };
    dispatch(removeItems(item));
  };

  return (
    <Section>
      <ImgWrapper>
        <img
          src={`./assets/img/pizza${id}.png`}
          alt="Pizza"
          width="80"
          height="80"
        />
      </ImgWrapper>
      <InfoWrapper>
        <h3>{name}</h3>
        <p>
          {type} dough, {size} inch
        </p>
      </InfoWrapper>
      <Counter>
        <ButtonCircle
          disabled={count === 1}
          onClick={onClickRemove}
          aria-label="minus"
        >
          <CountIcon as={BsDashLg} />
        </ButtonCircle>
        <span>{count}</span>
        <ButtonCircle onClick={onClickAdd} aria-label="plus">
          <CountIcon as={BsPlusLg} />
        </ButtonCircle>
      </Counter>
      <Price>
        <b>{price * count}$</b>
      </Price>
      <Remove>
        <ButtonCircle onClick={onClickDelete} remove aria-label="delete">
          <CountIcon as={BsPlusLg} />
        </ButtonCircle>
      </Remove>
    </Section>
  );
});
export default CartItem;
