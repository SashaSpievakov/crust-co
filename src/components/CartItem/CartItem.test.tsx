import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';

import CartItem from './CartItem';
import { ICartItem } from '../../models/ICartItem';
import renderWithStoreAndThemeProvider from '../../tests/helpers/renderWithStoreAndThemeProvider';

const CartItemProps: ICartItem = {
  id: '2',
  name: 'Pepperoni Pizza',
  price: 9,
  size: 16,
  type: 'traditional',
  count: 2,
};

describe('Cart Empty Test', () => {
  beforeEach(() => {
    renderWithStoreAndThemeProvider(<CartItem {...CartItemProps} />);
  });

  test('renders image', async () => {
    const image = screen.getByRole('img', { name: /pizza/i });
    expect(image).toBeInTheDocument();
  });

  test('renders heading', () => {
    const heading = screen.getByRole('heading', {
      name: /pepperoni pizza/i,
    });
    expect(heading).toBeInTheDocument();
  });

  test('renders description', () => {
    const description = screen.getByText(/traditional dough, 16 inch/i);
    expect(description).toBeInTheDocument();
  });

  test('renders price', () => {
    const price = screen.getByText(/18\$/i);
    expect(price).toBeInTheDocument();
  });

  describe('test buttons', () => {
    test('renders all buttons', () => {
      const buttons = screen.getAllByRole('button');
      expect(buttons[0]).toBeInTheDocument();
      expect(buttons[0]).toBeEnabled();
      expect(buttons[1]).toBeInTheDocument();
      expect(buttons[2]).toBeInTheDocument();
      expect(buttons[3]).toBeUndefined();
    });
  });
});
