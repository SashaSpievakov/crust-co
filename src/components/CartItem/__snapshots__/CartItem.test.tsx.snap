// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem Tests renders the CartItem component 1`] = `
Object {
  "children": Array [
    <div
      className="sc-bYMpWt bnFYVZ"
    >
      <picture>
        <source
          srcSet="./assets/img/pizza6.webp"
          type="image/webp"
        />
        <source
          srcSet="./assets/img/pizza6.png"
          type="image/png"
        />
        <img
          alt="Cheesburger Pizza pizza"
          height="80"
          src="./assets/img/pizza6.png"
          width="80"
        />
      </picture>
    </div>,
    <section
      className="sc-kMjNwy gJsxfE"
    >
      <h3>
        Cheesburger Pizza
      </h3>
      <p>
        thin
         dough, 
        14
         inch
      </p>
    </section>,
    <section
      className="sc-ilhmMj fPuPTA"
    >
      <button
        aria-label="minus"
        className="sc-hLBbgP sc-iBYQkv sc-ftTHYK fGqcTx iDltE YugEl"
        disabled={false}
        onClick={[Function]}
      >
        <svg
          className="sc-jIRcFI kihpWA"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 8a1 1 0 0 1 1-1h14a1 1 0 1 1 0 2H1a1 1 0 0 1-1-1z"
          />
        </svg>
      </button>
      <span>
        13
      </span>
      <button
        aria-label="plus"
        className="sc-hLBbgP sc-iBYQkv sc-ftTHYK fGqcTx iDltE YugEl"
        onClick={[Function]}
      >
        <svg
          className="sc-jIRcFI kihpWA"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
          />
        </svg>
      </button>
    </section>,
    <h3
      className="sc-hhOBVt kgouOf"
    >
      91
      $
    </h3>,
    <div
      className="sc-gYbzsP kNflxz"
    >
      <button
        aria-label="delete"
        className="sc-hLBbgP sc-iBYQkv sc-ftTHYK fGqcTx iDltE kTqzjG"
        onClick={[Function]}
      >
        <svg
          className="sc-jIRcFI kihpWA"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
          />
        </svg>
      </button>
    </div>,
  ],
  "props": Object {
    "className": "sc-ezOQGI bqqjuW",
  },
  "type": "article",
}
`;

exports[`CartItem Tests renders the CartItem component with a disabled button 1`] = `
Object {
  "children": Array [
    <div
      className="sc-bYMpWt bnFYVZ"
    >
      <picture>
        <source
          srcSet="./assets/img/pizza2.webp"
          type="image/webp"
        />
        <source
          srcSet="./assets/img/pizza2.png"
          type="image/png"
        />
        <img
          alt="Pepperoni Pizza pizza"
          height="80"
          src="./assets/img/pizza2.png"
          width="80"
        />
      </picture>
    </div>,
    <section
      className="sc-kMjNwy gJsxfE"
    >
      <h3>
        Pepperoni Pizza
      </h3>
      <p>
        traditional
         dough, 
        16
         inch
      </p>
    </section>,
    <section
      className="sc-ilhmMj fPuPTA"
    >
      <button
        aria-label="minus"
        className="sc-hLBbgP sc-iBYQkv sc-ftTHYK fGqcTx iDltE YugEl"
        disabled={true}
        onClick={[Function]}
      >
        <svg
          className="sc-jIRcFI kihpWA"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 8a1 1 0 0 1 1-1h14a1 1 0 1 1 0 2H1a1 1 0 0 1-1-1z"
          />
        </svg>
      </button>
      <span>
        1
      </span>
      <button
        aria-label="plus"
        className="sc-hLBbgP sc-iBYQkv sc-ftTHYK fGqcTx iDltE YugEl"
        onClick={[Function]}
      >
        <svg
          className="sc-jIRcFI kihpWA"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
          />
        </svg>
      </button>
    </section>,
    <h3
      className="sc-hhOBVt kgouOf"
    >
      9
      $
    </h3>,
    <div
      className="sc-gYbzsP kNflxz"
    >
      <button
        aria-label="delete"
        className="sc-hLBbgP sc-iBYQkv sc-ftTHYK fGqcTx iDltE kTqzjG"
        onClick={[Function]}
      >
        <svg
          className="sc-jIRcFI kihpWA"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
          />
        </svg>
      </button>
    </div>,
  ],
  "props": Object {
    "className": "sc-ezOQGI bqqjuW",
  },
  "type": "article",
}
`;
